### PR TITLE
[beta] Do not accidentally treat multi-segment meta-items as single-segment

### PR DESCRIPTION
**Pull request by petrochenkov**
Partial backport of https://github.com/rust-lang/rust/pull/58899 to beta

### CODE REVIEW

**Comment by pnkfelix** [comment]
Discussed at T-compiler meeting. We are inclined towards a lukewarm beta-accept

Why lukewarm:
 * it would have been nice if some of the extraneous changes were not part of the PR, just to make it more immediately clear on casual review what all the effects of this PR actually are.
 * it is not clear if any of the bugs fixed by this are ultra-high priority. Our understanding is that it is fixing a set of stable-to-stable regressions. While such regressions are of course bad, we do have a history of letting old stable-to-stable regressions lie unfixed, under the assumption that if the users have gotten by this long without a fix, they can wait the additional 6-weeks for the trains to roll the code out.

Having said that, we trust @petrochenkov wouldn't have put the work into making this PR if it wasn't actually important.

However, I also want to make sure this PR gets a proper review to make sure that the changes it makes are the right subset of #58899. 

So:
 * I am reassigning it to @estebank for review.
 * Once that review is done, then @petrochenkov can feel free to mark as beta-accepted.

**Comment by petrochenkov** [comment]
>it would have been nice if some of the extraneous changes were not part of the PR

The extraneous changes should be left in https://github.com/rust-lang/rust/pull/58899, everything here should be necessary due to removal of the old methods.
I've noticed the `expand.rs` change was mentioned on Zulip, but that's probably the only change that slipped through.

File: src/librustc_resolve/build_reduced_graph.rs [hunk 1]
```diff
@@ -815,9 +814,9 @@ impl<'a> Resolver<'a> {
                             break;
                         }
                         MetaItemKind::List(nested_metas) => for nested_meta in nested_metas {
-                            match nested_meta.word() {
-                                Some(word) => single_imports.push((word.name(), word.span)),
-                                None => ill_formed(nested_meta.span),
+                            match nested_meta.ident() {
+                                Some(ident) if nested_meta.is_word() => single_imports.push(ident),
```
Comment (estebank) [inline src/librustc_resolve/build_reduced_graph.rs]: For a follow up PR, I see this pattern in multiple places, it'd be nice to add a method `get_word_ident()` to `nested_meta` so you can use `if let` everywhere the alternative is discarded/an error.

**Pull request merged**